### ISSUE TITLE
Upgrade to Kubernetes 1.17

### ISSUE THREAD

**Issue by roberthbailey** [googleforgames/agones#1824]
We aren't ready for this yet, but now that we are wrapping up the upgrade to 1.16 we have some steps we want to proactively capture for 1.17 so it's time to start writing them down. 

List of items to do for upgrading to 1.17 (this was started from the list in the [1.16 issue](https://github.com/googleforgames/agones/issues/1649) and likley need further updates):

- [x] Update e2e cluster to run against 1.17
  - [x] Recreate cluster with new scripts (after updating the GKE terraform module)
  - [x] Update kubectl in `e2e-image/Dockerfile`
- [x] Update prow cluster to use 1.17 (even though we aren't using it yet, we should keep it in sync)
  - [x] Recreate cluster with new scripts
- [x] Update the dev tooling to create 1.17 clusters
  - [x] GKE
  - [x] Minikube
  - [x] Kind
  - [x] Update kubectl
- [x] Update terraform submodules
  - [x] GKE
  - [x] Azure
  - [x] EKS
- [x] Update documentation for creating clusters to 1.17
  - [x] Usage requirements
  - [x] GKE
  - [x] Minikube
  - [x] EKS
  - [x] AKS
- [x] Update links to k8s documentation
  - [x] examples/fleet.yaml
  - [x] examples/fleetautoscaler.yaml
  - [x] examples/gameserver.yaml
  - [x] site/content/en/docs/Reference/fleet.md
  - [x] site/content/en/docs/Reference/fleetautoscaler.md
  - [x] site/content/en/docs/Reference/gameserver.md
  - [x] site/content/en/docs/Advanced/limiting-resources.md
  - [x] site/assets/templates/crd-doc-config.json
- [x] Update client-go
- [x] Regenerate CRD Kubernetes client libraries
- [x] Move API references from beta to v1 (see the [Kubernetes v1.16 Release Notes](https://v1-16.docs.kubernetes.io/docs/setup/release/notes/#no-really-you-must-read-this-before-you-upgrade))
  - [x] CRDs
  - [x] Admission Webhooks

**Comment by markmandel** [comment]
We should target this at our 1.11.0 release - 1.17 is available across all three cloud providers.

**Comment by markmandel** [comment]
Wondering if we want to add running a long running version of #1867 to ensure there are no memory leaks in client-go when we upgrade?

This is the second time (#1871) we've been bitten by a memory leak in client-go.

**Comment by roberthbailey** [comment]
https://github.com/googleforgames/agones/issues/1799 is captured as a checkbox in the list above but also in a separate issue.

**Comment by markmandel** [comment]
Current status: Moving the CRDs to v1.

**Comment by markmandel** [comment]
Going to tackle upgrading go-client - it likely would need to happen after we get some more CRD v1 stuff in, but I expect we can manage the merge, so I figured I can get the work running in parallel and handle the merging.

**Comment by markmandel** [comment]
Ran an overnight allocation load test.

`TESTRUNSCOUNT=30 ./runAllocation.sh 40 100`

I can see a minor increase in memory over time, but looking at the heap graph, I don't see any egregious memory usage, so I'm expecting that it is likely normal memory usage, that will plateau over time.
![image](https://user-images.githubusercontent.com/298370/101809589-960b6280-3acc-11eb-8da1-a9737f74093d.png)

No CPU leaks.

Heap Graph:
![image](https://user-images.githubusercontent.com/298370/101810004-1f229980-3acd-11eb-9048-4182b7c90957.png)

Usually when we have a memory leak, it's _extremely_ obvious because it takes up a majority of the memory space.

**Comment by markmandel** [comment]
All items have now been checked! Happy to close this issue if everyone else is!

**Comment by markmandel** [comment]
Running some additional allocations over the past few hours - and yeah, looks like it's plateau's around the 675MB mark (on alloc), so what we're seeing it just a gradual increase to that point. :sweat: 

![image](https://user-images.githubusercontent.com/298370/101842080-257c3a00-3afc-11eb-9198-23452e759b3a.png)

**Comment by markmandel** [comment]
Poking at this some more because I wasn't 100% happy - and it _seems_ like it's okay, as it keep coming down to that baseline.

(This is the past 23 hours, which includes 2 load tests)

![image](https://user-images.githubusercontent.com/298370/101853772-08079a00-3b15-11eb-9c39-9f1654620cc8.png)

Nothing in the heap aloc graph seems to stands out;

![image](https://user-images.githubusercontent.com/298370/101854752-fc1cd780-3b16-11eb-9869-35d6e2653097.png)

I do see:

```
32MB 14.32%                | k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc /go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/store.go:85
```
Which was smaller before, but that (a) could be coincidence and (b) there's nothing in that function that actually stores anything, it's all transient String creation.

Maybe I'm just being overtly worrisome. :shrug:

**Comment by markmandel** [comment]
I'm going to say let's close this issue, since RC is tomorrow. If it looks like something has gone wrong during RC week, we can reopen a new issue. Sound good?

**Comment by markmandel** [comment]
Closing, since no objections!
